File: platforms/ios/Sources/Feedback/Schema.swift
import Foundation

/**
 Swift mirror of the wire format in `packages/core`.

 Third hand-maintained copy of one schema. The drift risk is real and called out
 in the README — generating all three from a single source is the correct fix.

 Deliberately free of UIKit so schema tests run on any platform.
 */

public enum Platform: String, Codable {
    case web, ios, android
}

public enum ReportKind: String, Codable {
    case bug, feedback, question
}

public struct Reporter: Codable, Equatable {
    public var email: String?
    public var fullName: String?
    public var externalId: String?

    public init(email: String? = nil, fullName: String? = nil, externalId: String? = nil) {
        self.email = email
        self.fullName = fullName
        self.externalId = externalId
    }
}

public struct ScreenInfo: Codable, Equatable {
    public let width: Int
    public let height: Int
    public let pixelRatio: Double
}

public struct DeviceContext: Codable, Equatable {
    public var platform: Platform = .ios
    public var sdkVersion: String
    public var osName: String? = "iOS"
    public var osVersion: String?
    public var deviceModel: String?
    public var locale: String?
    public var timezone: String?
    public var screen: ScreenInfo?
    /// The screen/route the user was on. The native analogue of a URL.
    public var route: String?
    public var appVersion: String?
    public var appBuild: String?
    public var networkType: String?

    public init(sdkVersion: String) {
        self.sdkVersion = sdkVersion
    }
}

public struct LogEntry: Codable, Equatable {
    public let level: String
    public let message: String
    public let timestamp: Int64
}

public struct Attachment: Codable, Equatable {
    public let id: String
    public let kind: String
    public let mimeType: String
    public let byteSize: Int
    public var width: Int?
    public var height: Int?
}

/// Normalized 0..1 so annotations survive any rescale, exactly as on web.
public struct Point: Codable, Equatable {
    public let x: Double
    public let y: Double

    public init(x: Double, y: Double) {
        self.x = x
        self.y = y
    }
}

public enum Annotation: Equatable {
    case rect(origin: Point, width: Double, height: Double, color: String)
    case arrow(from: Point, to: Point, color: String)
    case pen(points: [Point], color: String, strokeWidth: Double)
    /// Must be burned into the image, never merely overlaid.
    case blur(origin: Point, width: Double, height: Double)
}

extension Annotation: Encodable {
    private enum CodingKeys: String, CodingKey {
        case type, origin, width, height, color, from, to, points, strokeWidth
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case let .rect(origin, width, height, color):
            try container.encode("rect", forKey: .type)
            try container.encode(origin, forKey: .origin)
            try container.encode(width, forKey: .width)
            try container.encode(height, forKey: .height)
            try container.encode(color, forKey: .color)
        case let .arrow(from, to, color):
            try container.encode("arrow", forKey: .type)
            try container.encode(from, forKey: .from)
            try container.encode(to, forKey: .to)
            try container.encode(color, forKey: .color)
        case let .pen(points, color, strokeWidth):
            try container.encode("pen", forKey: .type)
            try container.encode(points, forKey: .points)
            try container.encode(color, forKey: .color)
            try container.encode(strokeWidth, forKey: .strokeWidth)
        case let .blur(origin, width, height):
            try container.encode("blur", forKey: .type)
            try container.encode(origin, forKey: .origin)
            try container.encode(width, forKey: .width)
            try container.encode(height, forKey: .height)
        }
    }
}

public struct Answer: Codable, Equatable {
    public let questionId: String
    public let value: String

    public init(questionId: String, value: String) {
        self.questionId = questionId
        self.value = value
    }
}

public enum SubmissionPayload: Equatable {
    case bugReport(kind: ReportKind, title: String, description: String?, annotations: [Annotation], logs: [LogEntry])
    case researchResponse(studyId: String, answers: [Answer], completed: Bool, durationMs: Int64?)
}

extension SubmissionPayload: Encodable {
    private enum CodingKeys: String, CodingKey {
        case type, kind, title, description, annotations, consoleLogs
        case studyId, answers, completed, durationMs
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case let .bugReport(kind, title, description, annotations, logs):
            try container.encode("bug_report", forKey: .type)
            try container.encode(kind, forKey: .kind)
            try container.encode(title, forKey: .title)
            // encodeIfPresent, not encode: absent and null are different on the
            // wire, and the validator treats them differently.
            try container.encodeIfPresent(description, forKey: .description)
            try container.encode(annotations, forKey: .annotations)
            try container.encode(logs, forKey: .consoleLogs)
        case let .researchResponse(studyId, answers, completed, durationMs):
            try container.encode("research_response", forKey: .type)
            try container.encode(studyId, forKey: .studyId)
            try container.encode(answers, forKey: .answers)
            try container.encode(completed, forKey: .completed)
            try container.encodeIfPresent(durationMs, forKey: .durationMs)
        }
    }
}

public struct Submission: Encodable {
    public let id: String
    public let projectId: String
    public let payload: SubmissionPayload
    public let device: DeviceContext
    public var reporter: Reporter?
    public var attachments: [Attachment]
    public var customData: [String: String]?
    public var sessionId: String?
    public var consent: ConsentRecord?
    public let createdAt: Int64

    public init(
        id: String = UUID().uuidString.lowercased(),
        projectId: String,
        payload: SubmissionPayload,
        device: DeviceContext,
        reporter: Reporter? = nil,
        attachments: [Attachment] = [],
        customData: [String: String]? = nil,
        sessionId: String? = nil,
        consent: ConsentRecord? = nil,
        createdAt: Int64 = Int64(Date().timeIntervalSince1970 * 1000)
    ) {
        self.id = id
        self.projectId = projectId
        self.payload = payload
        self.device = device
        self.reporter = reporter
        self.attachments = attachments
        self.customData = customData
        self.sessionId = sessionId
        self.consent = consent
        self.createdAt = createdAt
    }

    /// Serialized form the outbox stores and replays verbatim.
    public func toJSONData() throws -> Data {
        let encoder = JSONEncoder()
        // Stable key order makes the wire format diffable against the other two
        // platforms when chasing schema drift.
        encoder.outputFormatting = [.sortedKeys]
        return try encoder.encode(self)
    }
}

File: platforms/android/signals/src/main/kotlin/com/susatest/signals/Schema.kt
package com.susatest.signals

import java.util.UUID

/**
 * Kotlin mirror of the TypeScript wire format in `packages/core`.
 *
 * This is a hand-maintained copy, which is a real risk: three independent
 * implementations of one schema drift silently, and the drift surfaces as a
 * malformed field in production rather than a compile error. Generating all three
 * from one source is the correct long-term fix — see README.
 */

enum class Platform(val wire: String) {
    WEB("web"),
    IOS("ios"),
    ANDROID("android")
}

enum class ReportKind(val wire: String) {
    BUG("bug"),
    FEEDBACK("feedback"),
    QUESTION("question")
}

data class Reporter(
    val email: String? = null,
    val fullName: String? = null,
    val externalId: String? = null
)

data class DeviceContext(
    val platform: Platform = Platform.ANDROID,
    val sdkVersion: String,
    val osName: String? = "Android",
    val osVersion: String? = null,
    val deviceModel: String? = null,
    val locale: String? = null,
    val timezone: String? = null,
    val screenWidth: Int? = null,
    val screenHeight: Int? = null,
    val pixelRatio: Float? = null,
    /** The screen/route the user was on. The native analogue of a URL. */
    val route: String? = null,
    val appVersion: String? = null,
    val appBuild: String? = null,
    val networkType: String? = null
)

data class LogEntry(val level: String, val message: String, val timestamp: Long)

data class Attachment(
    val id: String,
    val kind: String,
    val mimeType: String,
    val byteSize: Long,
    val width: Int? = null,
    val height: Int? = null
)

/** Normalized 0..1 so annotations survive any rescale, exactly as on web. */
data class Point(val x: Float, val y: Float)

sealed class Annotation {
    data class Rect(val origin: Point, val width: Float, val height: Float, val color: String) : Annotation()
    data class Arrow(val from: Point, val to: Point, val color: String) : Annotation()
    data class Pen(val points: List<Point>, val color: String, val strokeWidth: Float) : Annotation()
    /** Must be burned into the image, never merely overlaid. */
    data class Blur(val origin: Point, val width: Float, val height: Float) : Annotation()
}

sealed class SubmissionPayload {
    data class BugReport(
        val kind: ReportKind,
        val title: String,
        val description: String? = null,
        val annotations: List<Annotation> = emptyList(),
        val logs: List<LogEntry> = emptyList()
    ) : SubmissionPayload()

    data class ResearchResponse(
        val studyId: String,
        val answers: List<Answer>,
        val completed: Boolean,
        val durationMs: Long? = null
    ) : SubmissionPayload()
}

/**
 * One answer.
 *
 * `values` exists because multi-choice must serialize as a JSON array, matching
 * web. Comma-joining into `value` would look correct on the wire and then
 * aggregate to zero: the dashboard's countChoices matches exact option strings,
 * so "Payment failed, Confusing pricing" matches neither option.
 */
data class Answer(
    val questionId: String,
    val value: String? = null,
    val values: List<String>? = null
)

data class Submission(
    val id: String = UUID.randomUUID().toString(),
    val projectId: String,
    val payload: SubmissionPayload,
    val device: DeviceContext,
    val reporter: Reporter? = null,
    val attachments: List<Attachment> = emptyList(),
    val customData: Map<String, String> = emptyMap(),
    val sessionId: String? = null,
    val consent: ConsentRecord? = null,
    val createdAt: Long = System.currentTimeMillis()
) {
    fun toJson(): String = JsonWriter().obj {
        str("id", id)
        str("projectId", projectId)
        obj("payload") { writePayload(payload) }
        obj("device") { writeDevice(device) }
        reporter?.let { r ->
            obj("reporter") {
                str("email", r.email)
                str("fullName", r.fullName)
                str("externalId", r.externalId)
            }
        }
        array("attachments", attachments) { a ->
            str("id", a.id)
            str("kind", a.kind)
            str("mimeType", a.mimeType)
            num("byteSize", a.byteSize)
            num("width", a.width)
            num("height", a.height)
        }
        if (customData.isNotEmpty()) {
            obj("customData") { customData.forEach { (k, v) -> str(k, v) } }
        }
        str("sessionId", sessionId)
        consent?.let { c ->
            obj("consent") {
                strArray("scopes", c.scopes.map { it.wire })
                str("policyVersion", c.policyVersion)
                num("grantedAt", c.grantedAt)
                str("source", c.source)
            }
        }
        num("createdAt", createdAt)
    }.toString()
}

private fun JsonWriter.ObjectScope.writePayload(payload: SubmissionPayload) {
    when (payload) {
        is SubmissionPayload.BugReport -> {
            str("type", "bug_report")
            str("kind", payload.kind.wire)
            str("title", payload.title)
            str("description", payload.description)
            array("annotations", payload.annotations) { writeAnnotation(it) }
            array("consoleLogs", payload.logs) { log ->
                str("level", log.level)
                str("message", log.message)
                num("timestamp", log.timestamp)
            }
        }
        is SubmissionPayload.ResearchResponse -> {
            str("type", "research_response")
            str("studyId", payload.studyId)
            array("answers", payload.answers) { a ->
                str("questionId", a.questionId)
                if (a.values != null) strArray("value", a.values) else str("value", a.value)
            }
            bool("completed", payload.completed)
            num("durationMs", payload.durationMs)
        }
    }
}

private fun JsonWriter.ObjectScope.writeAnnotation(annotation: Annotation) {
    when (annotation) {
        is Annotation.Rect -> {
            str("type", "rect")
            obj("origin") { num("x", annotation.origin.x); num("y", annotation.origin.y) }
            num("width", annotation.width)
            num("height", annotation.height)
            str("color", annotation.color)
        }
        is Annotation.Arrow -> {
            str("type", "arrow")
            obj("from") { num("x", annotation.from.x); num("y", annotation.from.y) }
            obj("to") { num("x", annotation.to.x); num("y", annotation.to.y) }
            str("color", annotation.color)
        }
        is Annotation.Pen -> {
            str("type", "pen")
            array("points", annotation.points) { p -> num("x", p.x); num("y", p.y) }
            str("color", annotation.color)
            num("strokeWidth", annotation.strokeWidth)
        }
        is Annotation.Blur -> {
            str("type", "blur")
            obj("origin") { num("x", annotation.origin.x); num("y", annotation.origin.y) }
            num("width", annotation.width)
            num("height", annotation.height)
        }
    }
}

/** Shared by submissions and analytics batches, so one device shape exists. */
internal fun JsonWriter.ObjectScope.writeDeviceContext(device: DeviceContext) = writeDevice(device)

private fun JsonWriter.ObjectScope.writeDevice(device: DeviceContext) {
    str("platform", device.platform.wire)
    str("sdkVersion", device.sdkVersion)
    str("osName", device.osName)
    str("osVersion", device.osVersion)
    str("deviceModel", device.deviceModel)
    str("locale", device.locale)
    str("timezone", device.timezone)
    if (device.screenWidth != null && device.screenHeight != null) {
        obj("screen") {
            num("width", device.screenWidth)
            num("height", device.screenHeight)
            num("pixelRatio", device.pixelRatio ?: 1f)
        }
    }
    str("route", device.route)
    str("appVersion", device.appVersion)
    str("appBuild", device.appBuild)
    str("networkType", device.networkType)
}
